storage/umass/umass.c

```diff
@@ -324,173 +731,278 @@ static void umass_msgthr(void *arg)
 static umass_dev_t *umass_devAlloc(void)
 {
 	umass_dev_t *dev;
+	int rv;
 
-	if ((dev = malloc(sizeof(umass_dev_t))) == NULL) {
+	dev = malloc(sizeof(umass_dev_t));
+	if (dev == NULL) {
 		fprintf(stderr, "umass: Not enough memory\n");
 		return NULL;
 	}
 
-	/* Get next device number */
-	if (umass_common.devices == NULL)
-		dev->id = 0;
-	else
-		dev->id = umass_common.devices->prev->id + 1;
+	rv = mutexCreate(&dev->lock);
+	if (rv < 0) {
+		free(dev);
+		return NULL;
+	}
 
-	dev->fileId = umass_common.lastId++;
+	idtree_alloc(&umass_common.devices, &dev->node);
+	dev->fileId = idtree_id(&dev->node);
 
-	if (mutexCreate(&dev->lock)) {
+	rv = snprintf(dev->path, sizeof(dev->path), "/dev/umass%d", dev->fileId);
+	if (rv < 0) {
+		resourceDestroy(dev->lock);
 		free(dev);
 		return NULL;
 	}
 
-	snprintf(dev->path, sizeof(dev->path), "/dev/umass%d", dev->id);
-
 	return dev;
 }
 
 
-static int umass_handleInsertion(usb_devinfo_t *insertion)
+static int umass_mountRoot(umass_dev_t *dev)
 {
+#ifdef UMASS_MOUNT_EXT2
+	int err;
+	oid_t roid;
+
+	do {
+		err = umass_mountFromDev(dev, LIBEXT2_NAME, &roid);
+		if (err < 0) {
+			break;
+		}
+		err = portRegister(roid.port, "/", &roid);
+		if (err < 0) {
+			break;
+		}
+		err = EOK;
+	} while (0);
```

Comment:
Maybe just `return err` instead of `break` in a `while`? There is no cleanup code or something else `do {} while (0)` would be useful for?